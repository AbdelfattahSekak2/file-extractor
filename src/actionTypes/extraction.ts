export const GetExtraction = {
  FAILURE: "GET_EXTRACTION_FAILURE",
  SUCCESS: "GET_EXTRACTION_SUCCESS",
  REQUEST: "GET_EXTRACTION_REQUEST",
}

export const ClearExtractionState = {
  DEFAULT: "CLEAR_EXTRACTION_STATE_DEFAULT",
}
