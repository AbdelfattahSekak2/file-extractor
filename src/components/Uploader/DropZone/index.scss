@import "../../../styles/mixins";

.dropzone-container {
  cursor: pointer;
  margin: auto;
  height: 400px;
  width: 400px;
  background-color: #fff;
  border-radius: 50%;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
  font-size: 16px;
  i.icon {
    font-size: 8em !important;
    color: $mindee-red !important;
  }
  .file-input {
    display: none;
  }
}
