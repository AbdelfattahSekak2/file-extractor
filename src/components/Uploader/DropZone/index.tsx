
import React from "react";
import { Icon } from "semantic-ui-react";

import "./index.scss";

interface IProps {
  onFileAdded: (file: File) => void;
  setError: (message: string) => void;
}

const supportedFiles = [
  "jpeg",
  "jpg",
  "png",
  "pdf",
  "zip",
];

const Dropzone: React.FunctionComponent<IProps> = ({
  onFileAdded,
  setError,
}) => {
  const onChange = (e: any) => handleOnFileAdded(e.target.files[0]);

  const handleOnFileAdded = (file: File) => {
    if (supportedFiles.some((e: string) => file.type.includes(e))) {
      onFileAdded(file);
    } else {
      setError("The file extension is not supported");
    }
  };

  const onDrop = (e: any) => {
    e.preventDefault();
    handleOnFileAdded(e.dataTransfer.files[0]);
  };

  const onDragOver = (e: any) => e.preventDefault();
  return (
    <label
      className="dropzone-container"
      onDrop={onDrop}
      onDragOver={onDragOver}
      htmlFor="file"
    >
      <input
        id="file"
        className="file-input"
        type="file"
        onChange={onChange}
        accept={"." + supportedFiles.join(", .")}
      />
      <Icon name="cloud upload" size="huge" />
    </label>
  );
};

export default Dropzone;
