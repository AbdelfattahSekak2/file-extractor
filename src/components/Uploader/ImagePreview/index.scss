@import "../../../styles/mixins";

.image-preview {
  background: white;
  border-radius: $border-radius;
  height: 35vw;
  position: relative;
  canvas {
    border-radius: $border-radius;
    height: 100%;
    margin: auto;
  }
  i.icon {
    cursor: pointer;
    position: absolute;
    right: 2% !important;
    top: 2%;
    color: $dark-grey;
    &.supported {
      color: white !important;
    }
  }

  .not-supported {
    height: 100%;
    padding: 20px;
    width: 100%;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    p {
      font-size: 28px;
      font-weight: 500px;
      text-align: center;
      color: $dark-grey;
    }
    i.icon {
      color: $dark-grey;
      position: unset;
    }
  }
}

@media (max-width: 500px) {
  .image-preview {
    height: 100vw;
  }
}

@media (min-width: 500px) and (max-width: 1000px) {
  .image-preview {
    height: 60vw;
  }
}

@media (min-width: 1000px) and (max-width: 1500px) {
  .image-preview {
    height: 40vw;
  }
}
