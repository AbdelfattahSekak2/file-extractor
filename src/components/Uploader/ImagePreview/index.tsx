import React, { useEffect, useRef, Ref } from "react";
import cx from "classnames";
import { Icon } from "semantic-ui-react";

import "./index.scss";

import Extraction from "../../../interfaces/extraction";

import Canvas, { handleCanvas } from "./helpers";

interface IProps {
  file: File;
  handleDelete: () => void;
  extraction: Extraction;
  setError: (message: string) => void;
  supported: boolean
}

const ImagePreview: React.FunctionComponent<IProps> = ({
  handleDelete,
  file,
  extraction,
  setError,
  supported,
}) => {
  const canvasRef: Ref<HTMLCanvasElement> = useRef(null);

  useEffect(() => {
    if (
      extraction.predictions
      && extraction.predictions.length > 0
      && canvasRef.current
      && supported
    ) {
      const canvas = new Canvas(canvasRef.current);
      handleCanvas(canvas, extraction, file, setError);
    }
  }, [extraction, file, setError, supported])
  return <div className="image-preview">
    <Icon name="times" size="big" onClick={handleDelete} className={cx({ supported })} />
    {
      supported
        ? <canvas ref={canvasRef} id="canvas" />
        : <div className="not-supported">
          <Icon name="frown" size="huge" />
          <p>Sorry, we cannot render this file</p>
        </div>
    }
  </div>;
};

export default ImagePreview;

