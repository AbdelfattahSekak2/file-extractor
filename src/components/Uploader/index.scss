.uploader-container {
  align-self: center;
  margin: auto;
}
