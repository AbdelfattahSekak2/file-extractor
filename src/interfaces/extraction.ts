export interface Coordinates {
  coord_x1: number;
  coord_x2: number;
  coord_y1: number;
  coord_y2: number;
}

export interface Tax extends Coordinates {
  rate: number;
  amount: number;
  probability: number;
}

interface Date extends Coordinates {
  date_INTL: string,
  date_US: string,
  probability: 0.972,
}

interface Total extends Coordinates {
  amount: number,
  probability: number,
}

export interface Prediction {
  orientation: {
    degrees: string;
    probability: number;
  },
  category: {
    value: string;
    probability: number;
  },
  date: Date
  total: Total;
  taxes: Tax[]
}

export default interface Extraction {
  details: {
    call_id: string;
    computation_time: number;
  };
  predictions: Prediction[];
}
