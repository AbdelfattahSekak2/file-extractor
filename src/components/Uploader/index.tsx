import React, { useState } from "react";

import Dropzone from "./DropZone";
import ImagePreview from "./ImagePreview";

import "./index.scss";

import Extraction from "../../interfaces/extraction";
import { ApiState } from "../../interfaces/apiState";
import { GetExtraction } from "../../actionTypes/extraction";

interface IProps {
  getExtraction: (file: FormData) => void;
  clearExtractionState: () => void;
  extraction: Extraction;
  apiPendingRequests: ApiState[];
  setError: (message: string) => void;
}

const Uploader: React.FunctionComponent<IProps> = ({
  getExtraction,
  clearExtractionState,
  extraction,
  apiPendingRequests,
  setError,
}) => {
  const [file, setFile] = useState();

  const readFile = (file: File) => {
    setFile(file);
    const formData = new FormData();
    formData.append("file", file);
    getExtraction(formData);
  }

  const handleDelete = () => {
    clearExtractionState();
    URL.revokeObjectURL(file);
    setFile(undefined);
  }
  
  const loading = apiPendingRequests.some((item: ApiState) => item.type === GetExtraction.REQUEST);
  let content: React.ReactElement | null = null;
  if (!loading) {
    if (file && extraction) {
      content = <ImagePreview
        setError={setError}
        file={file}
        handleDelete={handleDelete}
        extraction={extraction}
        supported={!file.type.includes("zip") && !file.type.includes("pdf")}
      />
    } else {
      content = <Dropzone setError={setError} onFileAdded={readFile} />;
    }
  }
  return <div className="uploader-container">{content}</div>;
};

export default Uploader;

