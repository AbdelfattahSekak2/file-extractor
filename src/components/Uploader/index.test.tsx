import React from "react";
import { ShallowWrapper, shallow } from "enzyme";

import Uploader from ".";

import { extraction } from "../../mocks/extraction";
import { GetExtraction } from "../../actionTypes/extraction";
import { act } from "react-test-renderer";

describe("Uploader component", () => {
  window.URL.revokeObjectURL = jest.fn();
  const setError = jest.fn();
  const getExtraction = jest.fn();
  const clearExtractionState = jest.fn();
  const file = new File([""], "filename", { type: "application/pdf" });
  const apiPendingRequests = [{ type: GetExtraction.REQUEST }];
  const defaultProps = {
    setError,
    getExtraction,
    clearExtractionState,
    extraction,
    apiPendingRequests: []
  }
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<Uploader {...defaultProps} />)
  })

  it("should render correctly", () => {
    expect(wrapper).not.toBeNull();

    expect(wrapper).toMatchSnapshot();
  })

  it("should render correctly", () => {
    wrapper = shallow(<Uploader {...defaultProps} apiPendingRequests={apiPendingRequests} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should call getExtraction when onFileAdded is fired", () => {
    //@ts-ignore
    wrapper.find("Dropzone").props().onFileAdded(file)
    expect(getExtraction).toHaveBeenCalled();
  });

  it("should read file correctly", () => {
    wrapper = shallow(<Uploader {...defaultProps} />)
    act(() => {
      //@ts-ignore
      wrapper.find("Dropzone").props().onFileAdded(file);
    });

    expect(wrapper.find("ImagePreview").prop("file")).toEqual(file);

    expect(getExtraction).toHaveBeenCalled()
  });

  it("should detele file correctly", () => {
    act(() => {
      //@ts-ignore
      wrapper.find("Dropzone").props().onFileAdded(file);
    });
    // @ts-ignore
    wrapper.find("ImagePreview").props().handleDelete();

    expect(clearExtractionState).toHaveBeenCalled();

    expect(window.URL.revokeObjectURL).toHaveBeenCalled();

    expect(wrapper.find("Dropzone")).toBeTruthy();
  });
})
