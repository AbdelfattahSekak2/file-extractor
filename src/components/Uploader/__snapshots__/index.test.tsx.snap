// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Uploader component should render correctly 1`] = `
<div
  className="uploader-container"
>
  <Dropzone
    onFileAdded={[Function]}
    setError={[MockFunction]}
  />
</div>
`;

exports[`Uploader component should render correctly 2`] = `
<div
  className="uploader-container"
/>
`;
